### DISCUSSION TITLE
How Can I call hooks in Atom Effect

### DISCUSSION THREAD

**Discussion by Gumichocopengin8** [facebookexperimental/Recoil#2041] [Q&A] [answered]
I wanna call `useNavigate` and `useLocation` from `react-router-dom` in Atom Effect to change query parameter.
However, I'm not allowed to call hooks besides top of function. Is there way to call them?
I was thinking to call them in `useEffect` in custom hooks, but realized that these are called multiple times if custom hooks is called in multiple components. I don't want to update query parameter unnecessarily multiple times. Are there any good way to handle it?
Thanks

**Comment by drarmstr** [comment, marked as answer]
Hooks can only be called from the render functions of React components or other hooks.  They should not be called from callbacks, effects, or otherwise outside of a component render.  They cannot be conditionally called.   [`useRecoilCallback()`](https://recoiljs.org/docs/api-reference/core/useRecoilCallback) can be useful to create a hook which can call other hooks and define a callback which can get/set selectors/atoms.  The [`recoil-sync`](https://recoiljs.org/docs/recoil-sync/introduction) library can also support a pattern which [creates a store](https://recoiljs.org/docs/recoil-sync/implement-store#example-syncing-with-react-props) from a component that can use props or hooks.

**Reply by Gumichocopengin8** [reply]
Thank you for the information. Let me take a look at them.
